misk/src/main/kotlin/misk/eventrouter/RealEventRouter.kt

```diff
@@ -4,80 +4,170 @@ import com.google.common.cache.CacheBuilder
+    class LeaveCluster : Action()
+  }
 
-          remoteSubscribers.get(event.topic).forEach {
-            it.send(eventJsonAdapter.toJson(event))
-          }
-        }
-      }
+  private val webSocketListener = object : WebSocketListener() {
+    override fun onMessage(webSocket: WebSocket, text: String) {
+      enqueue(Action.OnMessage(webSocket, text))
     }
   }
 
   private val topicPeer = object : TopicPeer {
     override fun acceptWebSocket(webSocket: WebSocket): WebSocketListener = webSocketListener
 
     override fun clusterChanged(clusterSnapshot: ClusterSnapshot) {
-      this@RealEventRouter.clusterSnapshot = clusterSnapshot
-      initCluster()
+      if (hasClusterSnapshot.compareAndSet(false, true)) {
+        this@RealEventRouter.clusterSnapshot = clusterSnapshot
+        logger.debug { "[${clusterSnapshot.self}]: cluster changed: $clusterSnapshot" }
+        executor.submit({ drainQueue() })
+      } else {
+        enqueue(Action.ClusterChanged(clusterSnapshot))
+      }
     }
   }
 
-  /** We queue up subscribes when we don't have a cluster. */
-  private fun initCluster() {
-    while (subscriptionQueue.isNotEmpty()) {
-      val queuedSubscribe = subscriptionQueue.pop()
-      subscribe(queuedSubscribe)
-    }
+  internal fun drainQueue() {
+    if (!hasClusterSnapshot.get()) return
+
+    while (true) {
+      val action = actionQueue.poll() ?: return
 
-    while (publishQueue.isNotEmpty()) {
-      val (topic, event) = publishQueue.pop()
-      publish(topic, event)
+      logger.debug { "[${clusterSnapshot.self}]: processing action: $action" }
+      when (action) {
+        is Action.LeaveCluster -> clusterConnector.leaveCluster(topicPeer)
+
+        is Action.ClusterChanged -> {
+          val removedHosts = clusterSnapshot.hosts.minus(action.newSnapshot.hosts)
+          for (host in removedHosts) {
+            logger.debug { "[${clusterSnapshot.self}]: invalidating $host" }
+            hostToSocket.invalidate(host)
+          }
+
+          clusterSnapshot = action.newSnapshot
+        }
+
+        is Action.OnMessage -> {
+          val socketEvent = eventJsonAdapter.fromJson(action.text)!!
+          when (socketEvent) {
+            is SocketEvent.Event -> {
+              localSubscribers.get(socketEvent.topic).forEach {
+                (it.listener as Listener<String>).onEvent(it as Subscription<String>,
+                    socketEvent.message)
+              }
+
+              remoteSubscribers.get(socketEvent.topic).forEach {
+                it.send(action.text)
+              }
+            }
+
+            is SocketEvent.Subscribe -> {
+              remoteSubscribers.put(socketEvent.topic, action.webSocket)
+              action.webSocket.send(eventJsonAdapter.toJson(SocketEvent.Ack(socketEvent.topic)))
+            }
+
+            is SocketEvent.Ack -> {
+              localSubscribers.get(socketEvent.topic).forEach {
+                it.onOpen()
```

Comment:
for follow-up: should be on yet another thread so that callers can do work immediately without blocking the world?